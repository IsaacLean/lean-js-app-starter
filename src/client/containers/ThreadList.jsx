import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import Loading from '../components/Loading'
import Threads from '../components/Threads'
import { endFetch, fetchThreads, setErr } from '../actions'
import { getThreads } from '../reducers'
import { setDocTitle } from '../util'

class ThreadList extends Component {
  componentDidMount() {
    setDocTitle()
    this.props.fetchThreads()
      .catch(err => {
        if(!this.props.threads || (Array.isArray(this.props.threads) && this.props.threads.length === 0)) {
          this.props.setErr('read', err)
        } else {
          this.props.endFetch()
        }
      })
  }

  render() {
    if(this.props.isFetching && Array.isArray(this.props.threads) && this.props.threads.length === 0) {
      return <Loading />
    }

    if(Array.isArray(this.props.threads) && this.props.threads.length === 0) {
      return <div className="center">No threads found. Be the first to post one by clicking on &quot;Start a New Thread&quot; above!</div>
    }
    return <Threads threads={this.props.threads} />
  }
}

ThreadList.propTypes = {
  endFetch: PropTypes.func.isRequired,
  fetchThreads: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
  setErr: PropTypes.func.isRequired,
  threads: PropTypes.arrayOf(PropTypes.shape({
    _id: PropTypes.string.isRequired,
    createdAt: PropTypes.string.isRequired,
    subject: PropTypes.string,
    comment: PropTypes.string.isRequired,
    replies: PropTypes.arrayOf(PropTypes.shape({
      _id: PropTypes.string.isRequired,
      createdAt: PropTypes.string.isRequired,
      comment: PropTypes.string.isRequired
    }))
  }))
}

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  threads: getThreads(state)
})

const mapDispatchToProps = dispatch => bindActionCreators({
  endFetch,
  fetchThreads,
  setErr
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(ThreadList)
