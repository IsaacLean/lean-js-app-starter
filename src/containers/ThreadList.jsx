import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import Threads from '../components/Threads'
import { fetchThreads } from '../actions'
import { getThreads } from '../reducers'

class ThreadList extends Component {
  componentDidMount() {
    this.props.fetchThreads()
  }

  render() {
    return <Threads threads={this.props.threads} />
  }
}

ThreadList.propTypes = {
  fetchThreads: PropTypes.func,
  threads: PropTypes.array
}

const mapStateToProps = state => ({
  threads: getThreads(state)
})

const mapDispatchToProps = dispatch => bindActionCreators({ fetchThreads }, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(ThreadList)
