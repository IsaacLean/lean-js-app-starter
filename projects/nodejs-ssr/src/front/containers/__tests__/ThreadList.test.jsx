import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow } from 'enzyme'
import React from 'react'

import { ThreadList } from '../ThreadList'

Enzyme.configure({ adapter: new Adapter() })

describe('ThreadList', () => {
  it('renders properly', () => {
    const component = shallow(<ThreadList isFetching={true} threads={[]} />)
    expect(component).toMatchSnapshot()
  })
})
