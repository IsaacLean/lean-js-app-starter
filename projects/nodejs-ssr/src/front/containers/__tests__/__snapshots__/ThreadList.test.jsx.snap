// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadList renders properly 1`] = `<Loading />`;
